File: JavaSlideshow/app/src/main/java/com/mobappdev/javaslideshow/MainActivity.java
package com.mobappdev.javaslideshow;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import com.mobappdev.javaslideshow.Observer.ObservedObject;
import com.mobappdev.javaslideshow.service.Slideshow;
import com.mobappdev.javaslideshow.model.Slide;

import java.time.LocalDate;
import java.util.Observable;
import java.util.Observer;

public class MainActivity extends AppCompatActivity implements Observer {

    public int currentSlideId = 0;
    public MainActivity(int id) {
        this.currentSlideId = id;
    }

    public MainActivity(){
    }

    private Slideshow holidaySlides = Slideshow.getInstance();
    private int totalSlides = 0;
    private int maxSlides = 4;
    private ImageView imageView;
    SharedPref sharedP = new SharedPref();
    ObservedObject watched;
    MainActivity watcher;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        // create watched and watcher objects
        watched = new ObservedObject(0);
        // watcher object listens to object change
        watcher = new MainActivity(0);
        watched.addObserver(watcher);

        final Context context = getApplicationContext();
        totalSlides = holidaySlides.getTotalSlides();

        if (totalSlides != maxSlides) {
            addDemoSlides();
            totalSlides = holidaySlides.getTotalSlides();
        }

        currentSlideId = sharedP.getValueInt(context, "slideId");

        Slide firstSlide = holidaySlides.next(currentSlideId);
        updateUI(firstSlide);

        if(imageView != null){
            imageView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    showNextSlide();
                    sharedP.save(context, "slideId", currentSlideId);
                }
            });
        }

        Button shuffleButton = (Button) findViewById((R.id.shuffle));
        shuffleButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                holidaySlides.shuffle();
            }
        });

        final Button sortButton = (Button) findViewById(R.id.sortButton);
        sortButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                holidaySlides.sort();
                Slide slide = holidaySlides.next(currentSlideId);
                updateUI(slide);
                if (holidaySlides.sortDirection.equals("asc")) {
                    sortButton.setText("SORT ASC");
                } else {
                    sortButton.setText("SORT DESC");
                }
            }
        });
    }

    public void showNextSlide(){
        watched.setValue(currentSlideId);
        currentSlideId++;
        if (currentSlideId == totalSlides) {
            currentSlideId = 0;
        }

        Slide nextSlide = holidaySlides.next(currentSlideId);
        if(watched.hasChanged()){
            TextView slideNo = (TextView) findViewById(R.id.slideNo);
            slideNo.setText("Slide ID " + nextSlide.getId());
            updateUI(nextSlide);
        }
    }

    public void updateUI(Slide slide) {
        TextView title = (TextView) findViewById(R.id.slideTitle);
        TextView description = (TextView) findViewById(R.id.description);
        TextView timestamp = (TextView) findViewById(R.id.timestamp);
        TextView gps = (TextView) findViewById(R.id.gpsText);
        int imgResId = getResources().getIdentifier(slide.getImageName(), "drawable", getPackageName());
        imageView = (ImageView) findViewById(R.id.slideImage);

        title.setText(slide.getTitle());
        description.setText(slide.getDescription());
        timestamp.setText(slide.getTimestamp() + "");
        if (gps != null) {
            gps.setText(slide.getGps());
        }
        if (imageView != null) {
            imageView.setImageResource(imgResId);
        }
    }

    public void addDemoSlides() {
        holidaySlides.addSlide(new Slide(1,
                "More Coffee!!",
                "coffee",
                "Hawaii",
                LocalDate.parse("2019-07-24"),
                "This was my coffee."));
        holidaySlides.addSlide(new Slide(2,
                "Beautiful sunset",
                "sunset",
                "Hawaii",
                LocalDate.parse("2019-07-24"),
                "This was a beautiful sunset."));
        holidaySlides.addSlide(new Slide(3,
                "Enjoy the beach",
                "beach",
                "Hawaii",
                LocalDate.parse("2019-07-24"),
                "This was a nice beach."));
        holidaySlides.addSlide(new Slide(4,
                "Sand and more",
                "sand",
                "Hawaii",
                LocalDate.parse("2019-07-25"),
                "This sand was very hot."));
    }


    @Override
    public void update(Observable o, Object arg) {
    }
}

File: KotlinSlideshow/app/src/main/java/at/fhj/iit/slideshowgroupa/MainActivity.kt
package at.fhj.iit.slideshowgroupa

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.VibrationEffect
import android.os.Vibrator
import android.util.Log
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.appcompat.app.AppCompatDelegate
import at.fhj.iit.slideshowgroupa.model.Slide
import at.fhj.iit.slideshowgroupa.service.Slideshow
import java.time.LocalDate
import kotlin.properties.Delegates
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.*
import kotlinx.coroutines.Dispatchers.Main
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL


class MainActivity : AppCompatActivity() {
    //SensorEventListener

    private val holidaySlides = Slideshow
    private var slideImgView: ImageView? = null
    private var totalSlides = 0
    private var gpsText: TextView? = null
    private val maxSlides: Int = 4

    //observable
    private var currentSlideId: Int by Delegates.observable(0) { prop, oldSlideId, newSlideId ->
        var slideNumber = newSlideId
        slideNumber++
        findViewById<TextView>(R.id.slideNo).text = "Slide $slideNumber of $totalSlides"
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        totalSlides = holidaySlides.getTotalSlides()
        val sharedPreference = SharedPreference(this)

        if (totalSlides != maxSlides) {
            addSomeDemoSlides()
            totalSlides = holidaySlides.getTotalSlides()
        }

        currentSlideId = sharedPreference.getValueInt("slideId")
        val firstSlide = holidaySlides.next(currentSlideId)

        updateUIWithSlide(firstSlide)
        slideImgView?.setOnClickListener {
            showNextSlide()
            sharedPreference.save("slideId", currentSlideId)
        }

        val shuffleButton = findViewById<Button>(R.id.shuffle)
        shuffleButton.setOnClickListener {
            holidaySlides.shuffle()
        }

        val sortButton = findViewById<Button>(R.id.sortButton)
        sortButton.setOnClickListener {
            holidaySlides.sort()
            val slide = holidaySlides.next(currentSlideId)
            updateUIWithSlide(slide)
            if (holidaySlides.sortDirection == "asc") {
                sortButton.text = "SORT ASC"
            } else {
                sortButton.text = "SORT DESC"
            }
        }
    }

    private fun showNextSlide() {
        currentSlideId++
        if (currentSlideId == totalSlides) currentSlideId = 0 else currentSlideId
        val nextSlide = holidaySlides.next(currentSlideId)
        updateUIWithSlide(nextSlide)
    }

    private fun updateUIWithSlide(slide: Slide) {
        val slideTitle = findViewById<TextView>(R.id.slideTitle)
        val description = findViewById<TextView>(R.id.description)
        val timestamp = findViewById<TextView>(R.id.timestamp)

        slideTitle.text = slide.title
        slideImgView = findViewById<ImageView>(R.id.slideImage)
        val imgResId = resources.getIdentifier(
            slide.imageName,
            "drawable",
            packageName
        )
        slideImgView?.setImageResource(imgResId)
        description.text = slide.description
        gpsText = findViewById<TextView>(R.id.gpsText)
        gpsText?.text = slide.GPS
        timestamp.text = (slide.timestamp).toString()
    }


    private fun addSomeDemoSlides() {
        holidaySlides.addSlide(
            Slide(
                1,
                "More Coffee!!",
                "coffee",
                "Hawaii",
                LocalDate.parse("2019-07-29"),
                "This was my coffee."
            )
        )
        holidaySlides.addSlide(
            Slide(
                2,
                "Beautiful sunset",
                "sunset",
                "Hawaii",
                LocalDate.parse("2019-07-29"),
                "This was a beautiful sunset."
            )
        )
        holidaySlides.addSlide(
            Slide(
                3,
                "Enjoy the beach",
                "beach",
                "Hawaii",
                LocalDate.parse("2019-07-24"),
                "This was a nice beach."
            )
        )
        holidaySlides.addSlide(
            Slide(
                4,
                "Sand and more",
                "sand",
                "Hawaii",
                LocalDate.parse("2019-07-25"),
                "This sand was very hot."
            )
        )
    }
}
